portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -21,25 +21,81 @@
 import javax.servlet.http.HttpServletResponse;
 import java.util.ArrayList;
 import com.google.gson.*;
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.SortDirection;
 
 /** Servlet that returns some example content. TODO: modify this file to handle comments data */
 @WebServlet("/data")
 public class DataServlet extends HttpServlet {
 
-  ArrayList<String> comments = new ArrayList<String>();
-
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    int numComments;
+    try {
+      numComments = Integer.parseInt(request.getParameter("number-comments"));
+    } catch(Exception e) {
+      numComments = 5;
+      System.out.println(request.getParameter("number-comments"));
+      System.out.println(request.getParameter("authuser"));
+      System.out.println(request.getQueryString());
+    }
+    
+    Query query = new Query("Comment").addSort("timestamp", SortDirection.DESCENDING);
+
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    PreparedQuery results = datastore.prepare(query);
+
+    ArrayList<Comment> comments = new ArrayList<>();
+    for (Entity entity : results.asIterable()) {
+      if(comments.size() >= numComments) {
+        break;
+      }
+
+      long id = entity.getKey().getId();
+      String name = (String) entity.getProperty("name");
+      String message = (String) entity.getProperty("message");
+      long timestamp = (long) entity.getProperty("timestamp");
+
+      comments.add(new Comment(id, name, message, timestamp));
+    }
+    
     response.setContentType("text/html;");
     String json = new Gson().toJson(comments);
     response.getWriter().println(json);
   }
 
   @Override
   public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    comments.add(request.getParameter("message"));
+    String name = request.getParameter("name");
+    String message = request.getParameter("message");
+    long timestamp = System.currentTimeMillis();
```

Comment:
Suppose timestamp matters to you - you should retrieve it the first thing when you get a chance. Doesn't really matter here since there are only two lines of statements ahead, but will matter if you wait until 1 or 2 RPC calls before retrieving the timestamp.

Reply:
Very true, I will fix this now